File: cpp/Path-Sum-II/Path-Sum-II.cpp
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> res;
        if (root == NULL)
            return res;
        vector<int> path;
        dfs(root, res, path, sum);
        return res;
    }
    void dfs(TreeNode* root, vector<vector<int>> &res, vector<int> &path, int sum)
    {
        if (root == NULL)
            return;
        path.push_back(root->val);
        bool isLeaf = (root->left == NULL) && (root->right == NULL);
        if (sum == root->val && isLeaf)
        {
            res.push_back(path);
        }
        dfs(root->left, res, path, sum - root->val);
        dfs(root->right, res, path,sum - root->val);
        path.pop_back();
    }
};

File: python/Path-Sum-II/Path-Sum-II.py
class Solution:
    def __init__(self):
        self.res = []
        self.path = []
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        if root is None:
            return self.res
        self.dfs(root, self.path, sum)
        return self.res
    def dfs(self, root, path, sum):
        if root is None:
            return
        self.path.append(root.val)
        if sum == root.val and (root.left is None and root.right is None):
            self.res.append(self.path[:])
        self.dfs(root.left, self.path, sum - root.val)
        self.dfs(root.right, self.path, sum - root.val)
        self.path.pop()
